### ISSUE TITLE
nw.js builds break when using system js 0.19.4

### ISSUE THREAD

**Issue by NervosaX** [systemjs/systemjs#841]
Since upgrading from 0.19.3 to 0.19.4, our node webkit (now known as nw.js) application fails to work. It seems to be running our application in a node context rather than a browser context. 

This will likely break all [nw.js](http://nwjs.io/)/[electron](https://github.com/atom/electron) applications as soon as they upgrade.

Additionally, is it possible to pin a system js version in jspm? As far as I can tell I only have the option of installing the latest systemjs, which we can't use.

**Comment by NervosaX** [comment]
https://github.com/NervosaX/systemjs-nw-bug This will let you replicate the issue:

To install:
========
* npm install
* jspm install
* Point the nw.js (use 0.12.3) binary at the directory, example:
/path/to/nw .

You should get the following error:
```
[15190:1012/174300:INFO:CONSOLE(0)] "Uncaught (in promise) ReferenceError: file:///home/adam/Programming/work/foo/jspm_packages/npm/babel-core@5.8.25.js:2
}).apply(__cjsWrapper.exports, __cjsWrapper.args);
         ^
__cjsWrapper is not defined
        Evaluating file:///home/adam/Programming/work/foo/jspm_packages/npm/babel-core@5.8.25.js
        Error loading file:///home/adam/Programming/work/foo/jspm_packages/npm/babel-core@5.8.25.js
        Error loading file:///home/adam/Programming/work/foo/test.js", source: file:///home/adam/Programming/work/foo/index.html (0)

```

**Comment by guybedford** [comment]
Thanks for posting. This sounds like it is due to the execution change from https://github.com/systemjs/systemjs/issues/821.

Do you have any ideas why this execution method wouldn't work for these cases?

**Comment by NervosaX** [comment]
@guybedford Generally they occur because it's incorrectly detected as node. 

In commit https://github.com/systemjs/systemjs/commit/808c5cddd9625f56648e4713cc05598e03ab53dd  it will be going through this part of the statement:

`if (typeof require != 'undefined' && typeof process != 'undefined' && process.versions && process.versions.node) {`

and returning true. It might be as simple as doing the document check first - it looks like it's trying to go the node way before trying the browser way?

**Comment by guybedford** [comment]
The change there was exactly for JSDom because the browser execution method falls down, so we're giving preference to Node.

I'm just trying to understand why the Node execution breaks down in these environments though? Is it to do with the execution scoping do you think?

**Comment by NervosaX** [comment]
It basically loses access to window, document, etc. as soon as it goes through `System.import`, suggesting it's using node's require to load the file. Perhaps checking for node webkit specifically can help?

https://gist.github.com/videlais/c426ff8d40d2ca35790f

**Comment by guybedford** [comment]
Ok thanks that makes complete sense. So these environments don't use Node globals, which is probably sensible.

We can revert https://github.com/systemjs/systemjs/commit/808c5cddd9625f56648e4713cc05598e03ab53dd then but we still need a way to support jsdom in https://github.com/systemjs/systemjs/issues/821.

@dmatteo is there a jsdom-only detection we can use instead to only use this execution method in that case?

**Comment by dmatteo** [comment]
@guybedford AFAIK there's no perfect recipe to detect jsdom, as the instance can be created in a lot of different ways.

However, if we say that 
```
isNode = typeof require != 'undefined' && typeof process != 'undefined' && process.versions && process.versions.node;
```

And we check for `if (isBrowser && isNode)` we have a very high level of confidence that we are working in a node browser like `jsdom` or `cheerio`.

What do you think?

**Comment by guybedford** [comment]
@dmatteo the problem here is that electron would pass the same check, but wouldn't work given this execution mode.

I think it may be best to revert the previous change in the coming release, and then allow the execution mode to be altered via a configuration option so that you can use this for jsdom support. If you come up with a reliable detection or have alternative suggestions do let me know too.

**Comment by dmatteo** [comment]
@guybedford I see...

How about browser detection?
[bowser](https://github.com/ded/bowser) is capable of detecting quite a few browsers, and if it's none of these (but still `isBrowser && isNode === true`) we could conclude it's `jsdom`?

**Comment by guybedford** [comment]
The breaking change here has been reverted in https://github.com/systemjs/systemjs/commit/c2a3ea9efbaed4a12d2bbceee5bc00396ce8d08c. @dmatteo the suggested fix for https://github.com/systemjs/systemjs/issues/814 should actually also solve your issue here.

**Comment by dmatteo** [comment]
@guybedford thanks, we will try that /cc @Munter 

Looking forward to 0.19.5 then!

**Comment by NervosaX** [comment]
Thanks @guybedford @dmatteo for this

**Comment by guybedford** [comment]
Released in SystemJS 0.19.5.

**Comment by dmatteo** [comment]
:+1:
